sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java

```diff
@@ -89,53 +89,22 @@
-                        responseHeaders,
-                        bodyAsObject));
-                } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
-                    throw logger.logExceptionAsError(new RuntimeException("Failed to deserialize 4-parameter"
-                        + " response. ", e));
-                }
+                return constructResponse(constructor, FOUR_PARAM_ERROR, logger, httpRequest, responseStatusCode,
+                    responseHeaders, bodyAsObject);
             case 5:
-                return decodedResponse.getDecodedHeaders()
-                    .map((Function<Object, Response<?>>) decodedHeaders -> {
-                        try {
-                            return constructor.newInstance(httpRequest,
-                                responseStatusCode,
-                                responseHeaders,
-                                bodyAsObject,
-                                decodedHeaders);
-                        } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
-                            throw logger.logExceptionAsError(new RuntimeException("Failed to deserialize 5-parameter"
-                                + " response with decoded headers. ", e));
-                        }
-                    })
-                    .switchIfEmpty(Mono.defer((Supplier<Mono<Response<?>>>) () -> {
-                        try {
-                            return Mono.just(constructor.newInstance(httpRequest,
-                                responseStatusCode,
-                                responseHeaders,
-                                bodyAsObject,
-                                null));
-                        } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
-                            throw logger.logExceptionAsError(new RuntimeException(
-                                "Failed to deserialize 5-parameter response without decoded headers.", e));
-                        }
-                    }));
+                return constructResponse(constructor, FIVE_PARAM_ERROR, logger, httpRequest, responseStatusCode,
+                    responseHeaders, bodyAsObject, decodedResponse.getDecodedHeaders());
             default:
-                throw logger.logExceptionAsError(
-                    new IllegalStateException("Response constructor with expected parameters not found."));
+                return monoError(logger, new IllegalStateException(INVALID_PARAM_COUNT));
         }
     }
+
+    private static Mono<Response<?>> constructResponse(Constructor<? extends Response<?>> constructor,
+        String exceptionMessage, ClientLogger logger, Object... params) {
+        return Mono.<Response<?>>fromCallable(() -> constructor.newInstance(params))
+            .onErrorMap(ex -> logger.logExceptionAsError(new RuntimeException(exceptionMessage, ex)));
```

Comment:
I think it's better to not throw `RuntimeException` directly. It might be better to use one of the exceptions defined in azure-core. 
https://azure.github.io/azure-sdk/java_introduction.html#java-errors-exception-tree